File: prototype/rails/src/app/actions/carts/current_cart_test.rb
require "commerce_test_case"

class CurrentCartTest < CommerceTestCase
  def setup
    @current_cart = Carts::CurrentCart.new
  end

  def test_a_visitor_with_no_cart_gets_one
    shopper = anonymous_customer

    cart = @current_cart.call(customer: shopper)

    assert_equal shopper, cart.customer
    assert_empty cart.items
  end

  def test_it_returns_the_same_cart_on_a_second_call
    shopper = anonymous_customer

    assert_equal @current_cart.call(customer: shopper), @current_cart.call(customer: shopper)
  end

  def test_a_merged_customer_keeps_shopping_with_the_cart_holding_items
    shopper = customer
    cart_for(shopper)
    filled = cart_holding(shopper, listing(seller))

    assert_equal filled, @current_cart.call(customer: shopper)
  end
end

File: prototype/php/src/app/Actions/Cart/CurrentCartTest.php
<?php

declare(strict_types=1);

namespace App\Actions\Cart;

use App\Models\Cart;
use App\Models\CartItem;

it('gives a customer without a cart one', function (): void {
    $customer = $this->anonymousCustomer();

    $cart = app(CurrentCart::class)($customer);

    expect($cart->customer_id)->toBe($customer->id)
        ->and(Cart::count())->toBe(1);
});

it('returns the same cart twice', function (): void {
    $customer = $this->anonymousCustomer();

    $first = app(CurrentCart::class)($customer);
    $second = app(CurrentCart::class)($customer);

    expect($second->id)->toBe($first->id);
});

it('picks the cart holding the items after a merge', function (): void {
    $customer = $this->verifiedCustomer();
    $this->cartFor($customer);
    $filled = $this->cartFor($customer);
    CartItem::create([
        'cart_id' => $filled->id,
        'listing_id' => $this->listing($this->seller())->id,
        'quantity' => 1,
    ]);

    expect(app(CurrentCart::class)($customer)->id)->toBe($filled->id);
});
